### PR TITLE
JS: support html attribute concatenations with single quotes

### PR DESCRIPTION
**Pull request by erik-krogh**
I found this while looking into CVE-2017-1000006, but this PR does not get us closer to flagging CVE-2017-1000006. 

HTML supports both single and double quotes, but our `js/incomplete-html-attribute-sanitization` query only considered double quotes. 

[Here is a query run showing new HTML attribute concatenations we find](https://lgtm.com/query/104181514671767412/).

### CODE REVIEW

File: javascript/ql/test/query-tests/Security/CWE-116/IncompleteSanitization/IncompleteHtmlAttributeSanitization.expected [hunk 1]
```diff
@@ -35,11 +51,20 @@ edges
...
 | tst.js:244:9:244:33 | s().rep ... /g, '') | tst.js:244:9:244:33 | s().rep ... /g, '') | tst.js:244:9:244:33 | s().rep ... /g, '') | Cross-site scripting vulnerability as the output of $@ may contain quotes when it reaches this attribute definition. | tst.js:244:9:244:33 | s().rep ... /g, '') | this final HTML sanitizer step |
 | tst.js:249:9:249:33 | s().rep ... ]/g,'') | tst.js:249:9:249:33 | s().rep ... ]/g,'') | tst.js:249:9:249:33 | s().rep ... ]/g,'') | Cross-site scripting vulnerability as the output of $@ may contain quotes when it reaches this attribute definition. | tst.js:249:9:249:33 | s().rep ... ]/g,'') | this final HTML sanitizer step |
+| tst.js:250:9:250:33 | s().rep ... ]/g,'') | tst.js:250:9:250:33 | s().rep ... ]/g,'') | tst.js:250:9:250:33 | s().rep ... ]/g,'') | Cross-site scripting vulnerability as the output of $@ may contain quotes when it reaches this attribute definition. | tst.js:250:9:250:33 | s().rep ... ]/g,'') | this final HTML sanitizer step |
```
Comment (esbena) [inline javascript/ql/test/query-tests/Security/CWE-116/IncompleteSanitization/IncompleteHtmlAttributeSanitization.expected:62]:
```suggestion
| tst.js:250:9:250:33 | s().rep ... ]/g,'') | tst.js:250:9:250:33 | s().rep ... ]/g,'') | tst.js:250:9:250:33 | s().rep ... ]/g,'') | Cross-site scripting vulnerability as the output of $@ may contain single quotes when it reaches this attribute definition. | tst.js:250:9:250:33 | s().rep ... ]/g,'') | this final HTML sanitizer step |
```

Comment (esbena) [inline javascript/ql/test/query-tests/Security/CWE-116/IncompleteSanitization/IncompleteHtmlAttributeSanitization.expected:62]:
```suggestion
| tst.js:250:9:250:33 | s().rep ... ]/g,'') | tst.js:250:9:250:33 | s().rep ... ]/g,'') | tst.js:250:9:250:33 | s().rep ... ]/g,'') | Cross-site scripting vulnerability as the output of $@ may contain double quotes when it reaches this attribute definition. | tst.js:250:9:250:33 | s().rep ... ]/g,'') | this final HTML sanitizer step |
```

Comment (esbena) [inline javascript/ql/test/query-tests/Security/CWE-116/IncompleteSanitization/IncompleteHtmlAttributeSanitization.expected:62]: Can we support a more specialized message?

**Review by esbena — COMMENTED** [review]
Hmm, can we get a bit more certainty that this does not introduce a large class of FPs? 

I remember playing around with both of the following formulations in my original PR, so I suspect I had a reason to avoid them in the end (although that reason may simply be a cautionary introduction of a new query).

```ql
  rep = "\"" and result = "double quotes" or
  rep = "'" and result = "single quotes"	
```
and your
```ql
  rep = ["\"", "'"] and result = "quotes"
```

File: javascript/ql/test/query-tests/Security/CWE-116/IncompleteSanitization/IncompleteHtmlAttributeSanitization.expected (hunk 1 above)
Comment (erik-krogh) [reply javascript/ql/test/query-tests/Security/CWE-116/IncompleteSanitization/IncompleteHtmlAttributeSanitization.expected]: Now the message says `single quotes` or `double quotes`.   
Is that the kind if specialization you were looking for?

Comment (esbena) [reply javascript/ql/test/query-tests/Security/CWE-116/IncompleteSanitization/IncompleteHtmlAttributeSanitization.expected]: yep

**Comment by esbena** [comment]
I am happy to merge if you are confident in the FP rate.

**Comment by erik-krogh** [comment]
> I am happy to merge if you are confident in the FP rate.

I am pretty confident - but I'm running an evaluation anyway. 

The query run in the top message does reveal that the below is sometimes seen as a HTML attribute concatenation: 

```JavaScript
$("input[id='" + filter + "']")
```

While it is not technically a HTML attribute, it behaves very similar to an HTML attribute, so I don't think it's a problem if we end up flagging some of those.

**Review by esbena — APPROVED** [review]

**Comment by erik-krogh** [comment]
> > I am happy to merge if you are confident in the FP rate.
> 
> I am pretty confident - but I'm running an evaluation anyway.

And an [evaluation](https://github.com/dsp-testing/erik-krogh-BCQS/blob/auto/data/workflow/single-quote-eval-4/reports/alert-comparison.md) showed no difference in results.   
Which wasn't all that surprising. 

I'm merging.

**Pull request merged**